File: mobile/lib/features/owner/pricing/data/models/coupon.dart
import '../../../../../core/utils/money_format.dart';

/// Mirrors `App\Http\Resources\CouponResource`.
class Coupon {
  const Coupon({
    required this.id,
    required this.code,
    required this.name,
    this.description,
    required this.discountType,
    required this.discountValue,
    this.minimumBookingAmount,
    this.maximumDiscountAmount,
    this.startsAt,
    this.expiresAt,
    this.usageLimit,
    this.usageLimitPerCustomer,
    required this.usageCount,
    required this.isActive,
    required this.firstBookingOnly,
    this.serviceIds = const [],
    this.categoryIds = const [],
  });

  factory Coupon.fromJson(Map<String, dynamic> json) => Coupon(
    id: json['id'] as String,
    code: json['code'] as String,
    name: json['name'] as String,
    description: json['description'] as String?,
    discountType: json['discount_type'] as String,
    discountValue: num.parse(json['discount_value'].toString()),
    minimumBookingAmount: json['minimum_booking_amount'] != null ? num.parse(json['minimum_booking_amount'].toString()) : null,
    maximumDiscountAmount: json['maximum_discount_amount'] != null ? num.parse(json['maximum_discount_amount'].toString()) : null,
    startsAt: json['starts_at'] as String?,
    expiresAt: json['expires_at'] as String?,
    usageLimit: json['usage_limit'] as int?,
    usageLimitPerCustomer: json['usage_limit_per_customer'] as int?,
    usageCount: json['usage_count'] as int,
    isActive: json['is_active'] as bool,
    firstBookingOnly: json['first_booking_only'] as bool,
    serviceIds: json['service_ids'] is List ? (json['service_ids'] as List<dynamic>).map((e) => e.toString()).toList() : const [],
    categoryIds: json['category_ids'] is List ? (json['category_ids'] as List<dynamic>).map((e) => e.toString()).toList() : const [],
  );

  final String id;
  final String code;
  final String name;
  final String? description;
  final String discountType;
  final num discountValue;
  final num? minimumBookingAmount;
  final num? maximumDiscountAmount;
  final String? startsAt;
  final String? expiresAt;
  final int? usageLimit;
  final int? usageLimitPerCustomer;
  final int usageCount;
  final bool isActive;
  final bool firstBookingOnly;
  final List<String> serviceIds;
  final List<String> categoryIds;

  String get discountLabel => discountType == 'percentage' ? '${formatMoney(discountValue)}%' : '₹${formatMoney(discountValue)}';
}

File: app/Models/Coupon.php
<?php

namespace App\Models;

use App\Enums\CouponDiscountType;
use App\Models\Concerns\BelongsToTenant;
use Illuminate\Database\Eloquent\Concerns\HasUlids;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\BelongsToMany;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\SoftDeletes;

class Coupon extends Model
{
    use BelongsToTenant, HasUlids, SoftDeletes;

    protected $fillable = [
        'code', 'name', 'description', 'discount_type', 'discount_value',
        'minimum_booking_amount', 'maximum_discount_amount', 'starts_at', 'expires_at',
        'usage_limit', 'usage_limit_per_customer', 'usage_count', 'is_active', 'first_booking_only',
    ];

    protected function casts(): array
    {
        return [
            'discount_type' => CouponDiscountType::class,
            'discount_value' => 'decimal:2',
            'minimum_booking_amount' => 'decimal:2',
            'maximum_discount_amount' => 'decimal:2',
            'starts_at' => 'datetime',
            'expires_at' => 'datetime',
            'is_active' => 'boolean',
            'first_booking_only' => 'boolean',
        ];
    }

    /**
     * The one normalization rule every lookup/creation goes through —
     * uppercased and trimmed, so `WELCOME10`/`welcome10`/` Welcome10 ` are
     * always the same logical code. See "Coupon code" in
     * LOYALTY_MEMBERSHIP_COUPON_ARCHITECTURE.md.
     */
    public static function normalizeCode(string $code): string
    {
        return strtoupper(trim($code));
    }

    public function tenant(): BelongsTo
    {
        return $this->belongsTo(Tenant::class);
    }

    public function services(): BelongsToMany
    {
        return $this->belongsToMany(Service::class, 'coupon_services')->withTimestamps();
    }

    public function categories(): BelongsToMany
    {
        return $this->belongsToMany(ServiceCategory::class, 'coupon_categories', 'coupon_id', 'category_id')->withTimestamps();
    }

    public function usages(): HasMany
    {
        return $this->hasMany(CouponUsage::class);
    }

    /** True only when neither `services()` nor `categories()` have any rows — applies tenant-wide. */
    public function appliesToAllServices(): bool
    {
        return $this->relationLoaded('services') && $this->relationLoaded('categories')
            ? ($this->services->isEmpty() && $this->categories->isEmpty())
            : ($this->services()->count() === 0 && $this->categories()->count() === 0);
    }
}
